serve/mlc_serve/model/torch_model.py

```diff
@@ -0,0 +1,643 @@
+import time
+import os
+import tempfile
+import socket
+from typing import List, Tuple, Sequence
+from pathlib import Path
+from concurrent.futures import ThreadPoolExecutor
+
+import structlog
+
+import torch
+import torch.multiprocessing as multiprocessing
+
+from transformers import AutoConfig
+
+try:
+    from vllm.model_executor.layers.sampler import get_logits
+    from vllm.model_executor.models.llama import LlamaForCausalLM
+    from vllm.model_executor.models.qwen import QWenLMHeadModel
+    from vllm.model_executor.models.phi import PhiForCausalLM
+    from vllm.model_executor.models.mistral import MistralForCausalLM
+    from vllm.model_executor.models.mixtral import MixtralForCausalLM
+    from vllm.model_executor import InputMetadata, SamplingMetadata
+    from vllm.model_executor.parallel_utils.parallel_state import (
+        initialize_model_parallel,
+    )
+    import rpyc
+    from rpyc.utils.classic import obtain
+    from rpyc.utils.server import ThreadedServer
+    from rpyc.utils.factory import unix_connect
+
+    support_torch_model = True
+
+except ImportError:
+    support_torch_model = False
+
+
+from .base import get_hf_config
+from .paged_cache_manager import KVCacheInfo, CacheManager
+from .model_common import (
+    prepare_inputs,
+    get_num_cache_blocks,
+    sample_from_logits,
+)
+
+from ..engine import (
+    get_prompt_sequence_id,
+    MLCServeEngineConfig,
+)
+from ..engine.model_module import (
+    DecodeRequest,
+    PrefillRequest,
+    TextGenerationResult,
+    TextGenerator,
+    RequestType,
+)
+from .sampler import SamplingState
+
+
+LOG = structlog.stdlib.get_logger(__name__)
+
+
+def init_cache_blocks(head_size, num_layers, num_heads, block_size, num_gpu_blocks):
```

Comment:
is there any difference with tvm path regarding kv cache profiling & initialization?

Reply:
It's doing the same thing but using PyTorch API.